File: csharp/tests/TestComplexProperties.cs
using MyAirtable;
using Xunit;

namespace MyAirtable.Tests;

/// <summary>
/// CS5.6 — Complex field-type CRUD against the live base: attachments (URL-based, async
/// processing), collaborators (users), computed fields (auto-number, formulas, created-time/by),
/// and the read-only button. Parity with the Java/Kotlin/Swift/Rust complex-property suites.
/// Linked records are covered by CS6.2.
/// </summary>
public class TestComplexProperties
{
    private readonly Airtable _airtable = TestSetup.MakeAirtable();

    [Fact]
    public async Task AttachmentFieldRoundTripsViaUrl()
    {
        var primaryKey = TestSetup.PrimaryKey("Complex", "Attach");
        const string url =
            "https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png";

        var created = await _airtable.Primary.CreateAsync(
            new PrimaryModel
            {
                PrimaryKey = primaryKey,
                Attachment = new List<AirtableAttachment> { new() { Url = url } },
            }
        );
        var recordId = created.Id!;
        try
        {
            // Airtable processes attachments asynchronously — the fresh record may carry just the
            // URL, or be fully populated with id/size once processing completes.
            Assert.NotNull(created.Attachment);
            Assert.Single(created.Attachment!);

            // Poll until Airtable has enriched the attachment (id populated).
            var current = created;
            for (var i = 0; i < 10; i++)
            {
                var first = current.Attachment is { Count: > 0 } a ? a[0] : null;
                if (!string.IsNullOrEmpty(first?.Id))
                    break;
                await Task.Delay(2000);
                current = await _airtable.Primary.GetAsync(recordId);
            }
            var enriched = current.Attachment ?? new List<AirtableAttachment>();
            Assert.Single(enriched);
            Assert.False(string.IsNullOrEmpty(enriched[0].Url), "attachment url is non-empty");
        }
        finally
        {
            await TryDeleteAsync(recordId);
        }
    }

    [Fact]
    public async Task UserAndMultiUserFieldsRoundTrip()
    {
        var primaryKey = TestSetup.PrimaryKey("Complex", "Users");
        // Matches the user ID used by the Rust/Swift/Java tests in the shared base.
        const string userId = "usrnZ4k98m0Ipji4e";

        var created = await _airtable.Primary.CreateAsync(
            new PrimaryModel
            {
                PrimaryKey = primaryKey,
                User = new AirtableCollaborator { Id = userId },
                UserAllowMultiple = new List<AirtableCollaborator> { new() { Id = userId } },
            }
        );
        var recordId = created.Id!;
        try
        {
            Assert.NotNull(created.User);
            Assert.Equal(userId, created.User!.Id);
            Assert.NotNull(created.UserAllowMultiple);
            Assert.Single(created.UserAllowMultiple!);
            Assert.Equal(userId, created.UserAllowMultiple![0].Id);

            var fetched = await _airtable.Primary.GetAsync(recordId);
            Assert.NotNull(fetched.User);
            Assert.Equal(userId, fetched.User!.Id);
        }
        finally
        {
            await TryDeleteAsync(recordId);
        }
    }

    [Fact]
    public async Task ComputedFieldsPopulateOnCreateAndReadBack()
    {
        var primaryKey = TestSetup.PrimaryKey("Complex", "Computed");
        var created = await _airtable.Primary.CreateAsync(
            new PrimaryModel
            {
                PrimaryKey = primaryKey,
                NumberInt = 10.0,
                NumberFloat = 5.0,
            }
        );
        var recordId = created.Id!;
        try
        {
            // Server-owned: auto number assigned (Value variant); created time populated.
            Assert.IsType<MaybeSpecialOrError<long>.Value>(created.AutoNumber);
            Assert.NotNull(created.CreatedTime);
            // Dedicated createdTime field decodes as a populated Value.
            Assert.IsType<MaybeSpecialOrError<DateTimeOffset>.Value>(created.CreatedAtTime);
            // Formula(ID) reflects the record id.
            Assert.Equal(recordId, created.FormulaId!.ValueOrDefault);
            // Formula(Simple) = numberInt + numberFloat = 15.
            Assert.Equal(15.0, created.FormulaSimple!.ValueOrDefault);
            // Created-by is populated with the caller's user info.
            Assert.NotNull(created.CreatedBy);
            Assert.NotNull(created.CreatedBy!.ValueOrDefault);

            var fetched = await _airtable.Primary.GetAsync(recordId);
            Assert.Equal(recordId, fetched.FormulaId!.ValueOrDefault);
            Assert.Equal(15.0, fetched.FormulaSimple!.ValueOrDefault);
            Assert.Equal(created.AutoNumber, fetched.AutoNumber);
        }
        finally
        {
            await TryDeleteAsync(recordId);
        }
    }

    [Fact]
    public async Task DurationRoundTripsAsNumericSeconds()
    {
        // CS5.5 magnitude assertion: the duration wire unit is seconds, so a 90-minute
        // TimeSpan round-trips exactly (not truncated/scaled).
        var primaryKey = TestSetup.PrimaryKey("Complex", "Duration");
        var duration = TimeSpan.FromMinutes(90);
        var created = await _airtable.Primary.CreateAsync(
            new PrimaryModel { PrimaryKey = primaryKey, Duration = duration }
        );
        var recordId = created.Id!;
        try
        {
            Assert.Equal(duration, created.Duration);
            var fetched = await _airtable.Primary.GetAsync(recordId);
            Assert.Equal(duration, fetched.Duration);
            Assert.Equal(5400.0, fetched.Duration!.Value.TotalSeconds);
        }
        finally
        {
            await TryDeleteAsync(recordId);
        }
    }

    [Fact]
    public async Task ButtonFieldDecodesOnRead()
    {
        var primaryKey = TestSetup.PrimaryKey("Complex", "Button");
        var created = await _airtable.Primary.CreateAsync(
            new PrimaryModel { PrimaryKey = primaryKey }
        );
        var recordId = created.Id!;
        try
        {
            // The button field decodes (either populated or absent) without error.
            var fetched = await _airtable.Primary.GetAsync(recordId);
            Assert.Equal(recordId, fetched.Id);
            if (fetched.Button?.ValueOrDefault is { } button)
                Assert.NotNull(button.Label ?? button.Url ?? "");
        }
        finally
        {
            await TryDeleteAsync(recordId);
        }
    }

    private async Task TryDeleteAsync(string? recordId)
    {
        if (string.IsNullOrEmpty(recordId))
            return;
        try
        {
            await _airtable.Primary.DeleteAsync(recordId);
        }
        catch (AirtableException)
        {
            // best-effort cleanup
        }
    }
}

File: cpp/tests/test_complex_properties.cpp
// Complex field-type CRUD against the live base: attachments (URL-based, async
// processing), collaborators, computed fields, and the read-only button
// (C# TestComplexProperties parity, 5 cases). Linked records: test_linked_records.cpp.
#include <catch2/catch_test_macros.hpp>

#include <chrono>
#include <thread>

#include "test_setup.hpp"

using namespace myairtable;
using myairtable_tests::make_airtable;
using myairtable_tests::primary_key;

namespace {
void try_remove(Airtable& airtable, const std::string& record_id) {
    try {
        airtable.primary().delete_one(record_id);
    } catch (const AirtableException&) {
    }
}
} // namespace

TEST_CASE("complex: attachment field round trips via url", "[crud][complex]") {
    auto airtable = make_airtable();
    const auto pk = primary_key("Complex", "Attach");
    const std::string url =
        "https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png";

    auto created = airtable.primary().create_one(PrimaryModel{
        .primary_key = pk,
        .attachment = std::vector<AirtableAttachment>{AirtableAttachment{.url = url}},
    });
    const auto record_id = *created.id;
    try {
        // Airtable processes attachments asynchronously — the fresh record may
        // carry just the URL, or be fully populated once processing completes.
        REQUIRE(created.attachment.has_value());
        REQUIRE(created.attachment->size() == 1);

        // Poll until Airtable has enriched the attachment (id populated).
        auto current = created;
        for (int i = 0; i < 10; ++i) {
            if (current.attachment.has_value() && !current.attachment->empty() &&
                current.attachment->front().id.has_value() &&
                !current.attachment->front().id->empty()) {
                break;
            }
            std::this_thread::sleep_for(std::chrono::seconds(2));
            current = airtable.primary().get_one(record_id);
        }
        REQUIRE(current.attachment.has_value());
        REQUIRE(current.attachment->size() == 1);
        REQUIRE(current.attachment->front().url.has_value());
        REQUIRE_FALSE(current.attachment->front().url->empty());
    } catch (...) {
        try_remove(airtable, record_id);
        throw;
    }
    try_remove(airtable, record_id);
}

TEST_CASE("complex: user and multi-user fields round trip", "[crud][complex]") {
    auto airtable = make_airtable();
    const auto pk = primary_key("Complex", "Users");
    // Matches the user ID used by the other targets' tests in the shared base.
    const std::string user_id = "usrnZ4k98m0Ipji4e";

    auto created = airtable.primary().create_one(PrimaryModel{
        .primary_key = pk,
        .user = AirtableCollaborator{.id = user_id},
        .user_allow_multiple =
            std::vector<AirtableCollaborator>{AirtableCollaborator{.id = user_id}},
    });
    const auto record_id = *created.id;
    try {
        REQUIRE(created.user.has_value());
        REQUIRE(created.user->id == user_id);
        REQUIRE(created.user_allow_multiple.has_value());
        REQUIRE(created.user_allow_multiple->size() == 1);
        REQUIRE(created.user_allow_multiple->front().id == user_id);

        auto fetched = airtable.primary().get_one(record_id);
        REQUIRE(fetched.user.has_value());
        REQUIRE(fetched.user->id == user_id);
    } catch (...) {
        try_remove(airtable, record_id);
        throw;
    }
    try_remove(airtable, record_id);
}

TEST_CASE("complex: computed fields populate on create and read back", "[crud][complex]") {
    auto airtable = make_airtable();
    const auto pk = primary_key("Complex", "Computed");
    auto created = airtable.primary().create_one(PrimaryModel{
        .primary_key = pk,
        .number_float = 5.0,
        .number_int = 10.0,
    });
    const auto record_id = *created.id;
    try {
        // Server-owned: auto number assigned (value variant); created time populated.
        REQUIRE(created.auto_number.has_value());
        REQUIRE(created.auto_number->is_value());
        REQUIRE(created.created_time.has_value());
        // The dedicated createdTime FIELD decodes as a populated value.
        REQUIRE(created.created_at_time.has_value());
        REQUIRE(created.created_at_time->is_value());
        // Formula(ID) reflects the record id.
        REQUIRE(created.formula_id->value() == record_id);
        // Formula(Simple) = numberInt + numberFloat = 15.
        REQUIRE(created.formula_simple->value() == 15.0);
        // Created-by is populated with the caller's user info.
        REQUIRE(created.created_by.has_value());
        REQUIRE(created.created_by->value().has_value());

        auto fetched = airtable.primary().get_one(record_id);
        REQUIRE(fetched.formula_id->value() == record_id);
        REQUIRE(fetched.formula_simple->value() == 15.0);
        REQUIRE(fetched.auto_number == created.auto_number);
    } catch (...) {
        try_remove(airtable, record_id);
        throw;
    }
    try_remove(airtable, record_id);
}

TEST_CASE("complex: duration round trips as numeric seconds", "[crud][complex]") {
    // Magnitude assertion: the duration wire unit is seconds, so a 90-minute
    // duration round-trips exactly (not truncated/scaled).
    auto airtable = make_airtable();
    const auto pk = primary_key("Complex", "Duration");
    const Duration duration{90.0 * 60.0};
    auto created = airtable.primary().create_one(PrimaryModel{
        .primary_key = pk,
        .duration = duration,
    });
    const auto record_id = *created.id;
    try {
        REQUIRE(created.duration == duration);
        auto fetched = airtable.primary().get_one(record_id);
        REQUIRE(fetched.duration == duration);
        REQUIRE(fetched.duration->seconds == 5400.0);
    } catch (...) {
        try_remove(airtable, record_id);
        throw;
    }
    try_remove(airtable, record_id);
}

TEST_CASE("complex: button field decodes on read", "[crud][complex]") {
    auto airtable = make_airtable();
    const auto pk = primary_key("Complex", "Button");
    auto created = airtable.primary().create_one(PrimaryModel{.primary_key = pk});
    const auto record_id = *created.id;
    try {
        // The button field decodes (either populated or absent) without error.
        auto fetched = airtable.primary().get_one(record_id);
        REQUIRE(fetched.id == record_id);
        if (fetched.button.has_value() && fetched.button->value().has_value()) {
            const auto& button = *fetched.button->value();
            REQUIRE((button.label.has_value() || button.url.has_value()));
        }
    } catch (...) {
        try_remove(airtable, record_id);
        throw;
    }
    try_remove(airtable, record_id);
}
